docs/advanced_topics.rst

```diff
@@ -46,6 +46,24 @@ Prefetching queue depth
 The purpose of this functionality is to average the processing time between batches when the variation from batch to batch is high.
 DALI pipeline allows buffering one or more batches of data ahead. This becomes important when the data processing time between batches could vary a lot. Default prefetch depth is 2. The user can change this value using the ``prefetch_queue_depth`` pipeline argument. For example, if the variation is bigger then it is recommended to prefetch more ahead.
 
+Running DALI pipeline
+---------------------
+
+DALI provides a couple of ways to run pipeline:
+
+- simple `run` method, which runs the computations and returns the results
+- `schedule_run`, `share_outputs` and `release_outputs` with fine grade control on output buffer lifetime
+- build-in iterators for MXNet, PyTorch, and TensorFlow
+
+The first API - `run` method launches the DALI pipeline, issuing the prefetch iterations if necessary, waits until the first batch is ready and returns the resulting buffers. Buffers are marked as in-use till the next invocations of `run` method. In many cases, it is wasteful as data is usually copied out to the native framework tensors and after that, thy can be set free and returned to DALI free buffer pool.
+
+For such purpose `schedule_run`, `share_outputs` and `release_outputs` methods are provided. `schedule_run` asks DALI to run prepare the next batch of data, doing prefetching if necessary. If the execution mode is set to asynchronous then this call should immediately return without waiting for the results and another task can be executed in parallel. When the batch of data is needed from DALI `share_outputs` should be called to return the result buffer. If it is not ready yet, it will wait for it. Then data is ready to be consumed. When DALI buffers are no longer needed, because data was copied or already consumed, `release_outputs` should be called to recycle DALI buffers and return them to pool so next iteration of computations can use them.
```

Comment:
```suggestion
For such purpose `schedule_run`, `share_outputs` and `release_outputs` methods are provided. `schedule_run` asks DALI to prepare the next batch of data, doing prefetching if necessary. If the execution mode is set to asynchronous then this call should immediately return without waiting for the results and another task can be executed in parallel. When the batch of data is needed from DALI `share_outputs` should be called to obtain the result buffer. If it is not ready yet, it will wait until data is ready to be consumed. When DALI buffers are no longer needed by the user, because data was copied out or processed, `release_outputs` should be called to recycle DALI buffers and return them to pool so next iteration of computations can use them.
```

Reply:
Done

Reply:
```suggestion
The second API, consisting of `run_no_result`, `share_outputs` and `release_outputs` allows the user to explicitly manage the lifetime of the output buffers. The `run_no_result` method instructs DALI to prepare the next batch of data, doing prefetching if necessary. If the execution mode is set to asynchronous, this call returns immediately, without waiting for the results, so another task can be executed in parallel. The data batch can be requested from DALI by calling `share_outputs`, which returns the result buffer. If it is not ready yet, it will wait for it. The data is ready as soon as the `share_outputs` method returns. When DALI buffers are no longer needed, because data was copied or already consumed, `release_outputs` should be called to return DALI buffers for reuse in subsequent iterations.
```